tracing/tracing.go

```diff
@@ -54,6 +54,32 @@ func OverrideTraceResourceName(sourceCtx context.Context, newResourceName string
 	return nil
 }
 
+// ExecuteWithTrace wraps a callable function with a Datadog tracer span.
+// If the context.Context does not contain Datadog metadata, then
+// the tracer span will not be recorded and the callable function
+// will be executed as normal.
+//
+// Example:
+//
+//	execRes, execErr := ExecuteWithTrace[*model.MyModel](
+//	    ctx,
+//	    myAmazingFunctionToCall,
+//	    "GetMyModel.SQL",
+//	    "GetMyModel",
+//	)
+func ExecuteWithTrace[T any](ctx context.Context, callable func() (T, error), source, op string) (T, error) {
+	traceSpan, traceSpanErr := CreateNestedTrace(ctx, fmt.Sprintf("%s.%s", source, op), op)
+
+	execResp, execErr := callable()
+
+	// NOTE: close only if context was given with Datadog metadata.
+	if traceSpanErr == nil {
+		traceSpan.Finish(tracer.WithError(execErr))
+	}
+
+	return execResp, execErr
```

Comment:
The `callable` will have no information about the new span. Ideally the span should be added to the context.Context, and then forwarded to the `callable` func.

For common external connections (like HTTP calls, Google Cloud SDK, Database-connections, gRPC calls etc) we should strive to add good wrappers that handles this automatically (extraction of existing spans from context.Context, creating child-spans, appending the child-spans to HTTP-headers/gRPC-metadata for distributed tracing).
Wrapper for HTTP client here: https://github.com/coopnorge/go-datadog-lib/blob/main/middleware/http/client.go
Wrapper for gRPC client here: https://github.com/coopnorge/go-datadog-lib/blob/main/middleware/grpc/client.go

Full list of already supported contribs in Datadog's lib: https://github.com/DataDog/dd-trace-go/tree/main/contrib

Reply:
Not sure if I got you guys but:

1. `callable` is a lambda function that controls outside packages - in conreceet project/microservice. By addeting context in lampda you forcing consumer to add this extra argument and that could be repetitive code as example:
```go
func (r *Repository) GetFooBar(ctx context.Context, id uint) (MyModel, error) {
	getMyModel := func() (*MyModel, error) {
		return r.GetMyModel(id)
	}

	execRes, execErr := coopDDTrace.ExecuteWithTrace[*MyModel](ctx, getMyModel, "GetMyModel.SQL", "GetMyModel")
	
	return execRes, execErr
}
```

2. `context` already must have information of datadog from middleware and it's the consumer's (programmer) responsibility to give correct context that it could create span from it. As package it check if it's correct on if not then tracing will be not enabled.

5. Of course tracing adds leniency but it's up to the consumer(programmer) to consider this action as maintainers of the library (package) you should not care or think about how it's used other wise it will be not flexible to be re-used as code.

Reply:
@bendiknesbo @arunpoudel  ^

Reply:
By adding the new nested span to the context, then any further spans will be able to be created as nested spans of the nested span. 

Take the following scenario:
1. The API receives an incoming HTTP/gRPC request, and the middleware creates a new Trace ( lets call it span "a") for the request, and adds this trace-information to the request's context.
2. The API then wants to trace a specific set of operations, so it calls `ExecuteWithTrace` (Creates span "b", as a child span of span "a").
3. One of the things we want to do within `ExecuteWithTrace`, is to call an external HTTP service. If we use the wrapped HTTP client from https://github.com/coopnorge/go-datadog-lib/blob/main/middleware/http/client.go, then the wrapped client will create a third span "c", but since `callable` as implemented in this PR does not have any reference to the span "b" (neither explicitly or via context).

This will create a hierarchy that looks like this:
```
.
└── a
    ├── b
    └── c
```

However, if we add the span `b` to a context, and pass that context to the `callable` func, then we will get the following hierarchy:
```
.
└── a
    └── b
        └── c
```

Reply:
If your callable code does not care about context, then thats fine. But if it cares about the context, then it would be nice to propagate the spans correctly.

```
func (r *Repository) GetFooBar(ctx context.Context, id uint) (MyModel, error) {
	getMyModel := func(_ context.Context) (*MyModel, error) {
		return r.GetMyModel(id) // Does not care about ctx
	}

	execRes, execErr := coopDDTrace.ExecuteWithTrace[*MyModel](ctx, getMyModel, "GetMyModel.SQL", "GetMyModel")
	
	return execRes, execErr
}

func (r *Repository) GetBaz(ctx context.Context, id uint) (MyModel, error) {
	getMyModel := func(ctx context.Context) (*MyModel, error) {
		return r.GetMyModel(ctx, id) // Cares about ctx
	}

	execRes, execErr := coopDDTrace.ExecuteWithTrace[*MyModel](ctx, getMyModel, "GetMyModel.SQL", "GetMyModel")
	
	return execRes, execErr
}

Reply:
I agree with @bendiknesbo here

Reply:
Anyway, add what you like on top, right now there is no generic function to wrap any execution. So please improve it if you have time, from my side I don't have capacity to do at the moment

Reply:
https://github.com/coopnorge/go-datadog-lib/pull/194

Reply:
I like @bendiknesbo 's option better. But do we really need this function? Can we not make `tracing/tracing.go`'s `getSpanFromContext` public and help the developer to trace these kind of stuffs?

We will probably have to make sure that we return a `noopSpan` which will minimize the nil checking the dev will have to do on their end.